WeatherDataFetcher.py

```diff
@@ -1,101 +1,166 @@
 import os
+import sys
 import datetime
 from typing import Tuple
 from geopy.geocoders import Nominatim
-import urllib.request, json
+import json
+import dotenv
+import requests
 
 
-class Weather_Forecast_Fetcher:
+def fetch_weather_data(api_key: str, location_coords: Tuple[int, int]) -> None:
     """
-    This class provides 3 days of forecast from an open API for a given location
-    api_key_open_weather_map.dat should contain the api key
+    Fetches weather data for 25 timepoints from the url if no recent data is available
+    Stores the data in the folder where it is executed for future use.
 
     Parameters:
-    ----------
-    city : stry
-      City for which the forcast will be given
-    location_coords : Tuple
-      Latitude and Longitude Coordinates of the location of interest (self.city)
-    api_key : str
-      API Key used for OpenWeatherMap
+    ---------
+    location_coords : Tuple[int:int]
+      (lat,long) coordinates for the city entered
+    recent_data_available : bool
+      if true, use cached values (load from file)
+      if false, reload from url (requests)
 
+    Returns: None
+    """
+
+    response = requests.get(
+        f"https://api.openweathermap.org/data/2.5/forecast?",
+        params={
+            "lat": location_coords[0],
+            "lon": location_coords[1],
+            "dt": 25,
+            "units": "metric",
+            "appid": api_key,
+        },
+    )
+
+    # Check if fetch_data_from_url is necessary
+    file_path_json = os.getcwd() + "/data.json"
+    fetch_data_from_url = False
+    if not os.path.exists(file_path_json):
+        fetch_data_from_url = True
+    else:
+        with open(file_path_json, "r", encoding="utf-8") as f:
+            data = json.loads(f.read())
+            now_utc_in_s = datetime.datetime.now(datetime.timezone.utc).timestamp()
+            oldest_utc_in_s = data["list"][0]["dt"]
+            # Data is more than day old - fetch_data_from_url required
+            if now_utc_in_s - oldest_utc_in_s > 86400:
+                fetch_data_from_url = True
+            # Different location is requested - fetch_data_from_url required
+            lat = data["city"]["coord"]["lat"]
+            lon = data["city"]["coord"]["lon"]
+            threshold = 0.001
+            if abs(lat - location_coords[0]) > threshold:
+                fetch_data_from_url = True
+            if abs(lon - location_coords[1]) > threshold:
+                fetch_data_from_url = True
+
+    if not os.path.exists(file_path_json) or fetch_data_from_url:
+        print("Getting new data from URL", file=sys.stderr)
+        data = response.json()
+        with open(file_path_json, "w", encoding="utf-8") as f:
+            json.dump(data, f, ensure_ascii=False, indent=4)
+    else:
+        print("Getting new data local file", file=sys.stderr)
+        with open(file_path_json, "r", encoding="utf-8") as f:
```

Comment:
Maybe find a way to load the data only one. Up you did it conditionally already. You could use as type `Optional[dict]`. Basically set it `None` and if it is `None` then fetch it.

Reply:
We can discuss this maybe quickly :innocent: